packages/explorer/src/app/(explorer)/[chainName]/worlds/[worldAddress]/explore/TablesViewer.tsx

```diff
@@ -33,12 +46,47 @@ type Props = {
   isLiveQuery: boolean;
 };
 
-export function TablesViewer({ table, query, isLiveQuery }: Props) {
+export function TablesViewer({ table, isLiveQuery }: Props) {
   const { id: chainId } = useChain();
   const indexer = indexerForChainId(chainId);
-  const { data: tableData, isPending, isError, error } = useTableDataQuery({ table, query, isLiveQuery });
+  const [query, setQuery] = useQueryState("query", parseAsString.withDefault(""));
   const [globalFilter, setGlobalFilter] = useQueryState("filter", parseAsString.withDefault(""));
   const [sorting, setSorting] = useQueryState("sort", parseAsJson<SortingState>().withDefault(initialSortingState));
+  const [page, setPage] = useQueryState("page", parseAsInteger.withDefault(0));
+  const [pageSize, setPageSize] = useQueryState("pageSize", parseAsInteger.withDefault(10));
+  const { data: tableData, isPending, isFetching, isError, error } = useTableDataQuery({ table, query, isLiveQuery });
+  const isLoading = isPending || (isFetching && !isLiveQuery);
+  const [pagination, setPagination] = useState<PaginationState>({
+    pageIndex: page,
+    pageSize,
+  });
+
+  // Sync pagination state when URL params change
+  useEffect(() => {
+    setPagination(() => ({
+      pageIndex: page,
+      pageSize: pageSize,
+    }));
+  }, [page, pageSize]);
+
+  const handlePaginationChange: OnChangeFn<PaginationState> = useCallback(
+    (updater) => {
+      const newPaginationState = typeof updater === "function" ? updater(pagination) : updater;
+      const { pageIndex: newPageIndex, pageSize: newPageSize } = newPaginationState;
+      setPage(newPageIndex);
+      setPageSize(newPageSize);
+
+      const decodedQuery = decodeURIComponent(query);
```

Comment:
the thing we use for query params doesn't give it back to use decoded already?

Reply:
answered below